### PR TITLE
Pig Latin transform

### PR DESCRIPTION
**Pull request by nick11roberts**
This transformation translates the original text into pig latin. Resulting sentences retain the same meaning as the original sentences, and are modified only on the word level. While the resulting sentences can still be interpreted by an English-speaking human annotator with some effort, the transformation yields sentences which are completely different on the word level from the original sentence. 

@vinayprabhu @rteehas @jjangsangy

### CODE REVIEW

File: transformations/pig_latin/transformation.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+        TaskType.TEXT_TAGGING,
+    ]
+    languages = ["en"]
+
+    def __init__(self, seed=0, max_outputs=1):
+        super().__init__(seed, max_outputs=max_outputs)
+
+    def generate(self, sentence: str):
```
Comment (csinva) [inline transformations/pig_latin/transformation.py:17]: It doesn't seem that max_outputs is actually used?

**Review by csinva — APPROVED** [review]
Hi @**[nick11roberts](https://github.com/nick11roberts)**, thanks for this interesting contribution! I'm a reviewer assigned to this PR. I've left a couple comments on the code. Here's my general review:

**Correctness**: All checks have passed - please pull "main" once into your branch.
**Interface**: I think you've chosen the correct interface. You might consider changing each word with some probability $p$ passed to the constructor rather than changing every word.
**Specificity**: The change is specific to anything which takes in text-level inputs.
**Novelty**: Different from other transformations in NL-Augmenter.
**Adding New Libraries**: Added the piglating library
**Test Cases**: 5 test cases added.
**Evaluating Robustness**: Robustness evaluation is not yet conducted.

File: transformations/pig_latin/transformation.py (hunk 1 above)
Comment (nick11roberts) [reply transformations/pig_latin/transformation.py:19]: Thanks for catching this. Per your suggestion, I added a per-word translation probability. Now that the output is non-deterministic for `0 < replace_prob < 1`, `max_outputs` is actually meaningful, so I went ahead and added functionality to produce multiple outputs.

**Comment by nick11roberts** [comment]
Thank you for the comments! I've fetched the current upstream commits and included your suggestion for using a translation probability. Accordingly, the transformation now meaningfully uses `max_outputs`.

**Comment by csinva** [comment]
Thanks for the additions, looks good to me!

**Comment by nick11roberts** [comment]
Thanks @csinva! I've also added evaluation results to the readme. Using the same evaluation procedure as shown in [the leaderboard](https://github.com/GEM-benchmark/NL-Augmenter/tree/main/evaluation), text classification models are substantially less robust to the PigLatin transform than all of the leading transformations for evaluating robustness. 

| Transformation              | roberta-base-SST-2   | bert-base-uncased-QQP   | roberta-large-mnli   | roberta-base-imdb   |
|:----------------------------|:---------------------|:------------------------|:---------------------|:--------------------|
| **PigLatin**       | **94.0->60.0 (-34.0)**   | **92.0->68.0 (-24.0)**      | **91.0->50.0 (-41.0)**   | **95.0->32.0 (-63.0)**  |
| BackTranslation             | 94.0->91.0 (-3.0)    | 92.0->90.0 (-2.0)       | 91.0->87.0 (-4.0)    | 95.0->92.0 (-3.0)   |
| ButterFingersPerturbation   | 94.0->89.0 (-5.0)    | 92.0->89.0 (-3.0)       | 91.0->88.0 (-3.0)    | 95.0->93.0 (-2.0)   |
| ChangePersonNamedEntities   | 94.0->94.0 (0.0)     | 92.0->92.0 (0.0)        | 91.0->89.0 (-2.0)    | 95.0->95.0 (0.0)    |
| CloseHomophonesSwap         | 94.0->91.0 (-3.0)    | 92.0->88.0 (-4.0)       | 91.0->89.0 (-2.0)    | 95.0->96.0 (1.0)    |
| DiscourseMarkerSubstitution | 94.0->94.0 (0.0)     | 92.0->92.0 (0.0)        | 91.0->91.0 (0.0)     | 95.0->95.0 (0.0)    |
| MixedLanguagePerturbation   | 94.0->90.0 (-4.0)    | 92.0->86.0 (-6.0)       | 91.0->86.0 (-5.0)    | 95.0->91.0 (-4.0)   |
| PunctuationWithRules        | 94.0->94.0 (0.0)     | 92.0->92.0 (0.0)        | 91.0->91.0 (0.0)     | 95.0->90.0 (-5.0)   |
| ReplaceNumericalValues      | 94.0->94.0 (0.0)     | 92.0->92.0 (0.0)        | 91.0->90.0 (-1.0)    | 95.0->95.0 (0.0)    |
| SentenceReordering          | 94.0->95.0 (1.0)     | 92.0->93.0 (1.0)        | nan                  | 95.0->94.0 (-1.0)   |

File: transformations/pig_latin/transformation.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+from tasks.TaskTypes import TaskType
+
+class PigLatin(SentenceOperation):
+    tasks = [
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+        TaskType.TEXT_TAGGING,
+    ]
+    languages = ["en"]
+
```
Comment (afnanmmir) [inline transformations/pig_latin/transformation.py:14]: Make sure to add keywords variable to your class

**Review by afnanmmir — CHANGES_REQUESTED** [review]
Hi, I am a reviewer for this pull request! Here are my general comments:
-  Correctness: Everything is written in python and black formatted, and all test cases pass
- Interface: I believe the sentence interface is the correct interface to use, as it is taking input text and returning a transformed version
- Specificity: Isn't necessarily specific, but does preserve information from text.
- Novelty: Is an original transformation that is a valid because it preserves information of the text.
- New Libraries: Add piglatin library.
- README: Detailed README that clearly explains the transformation and why it is important. Also gives robustness evaluation results.
- Data and Code Source: None added, but none needed
-  Applicable Tasks & Keywords: No keywords variable added
-  Test cases: 5 added.
-  Languages: English
Only thing to add is keyword variable, then your good to go!

**Comment by kaustubhdhole** [comment]
Thank you very much for the changes! This PR looks great and is a very important addition to NL-Augmenter. Also thanks to @csinva and @afnanmmir for the reviews! I am merging this now. @nick11roberts, please create a separate PR to add the keywords for your PR - I will merge it directly!

**Review by kaustubhdhole — APPROVED** [review]

**Comment by nick11roberts** [comment]
> Thank you very much for the changes! This PR looks great and is a very important addition to NL-Augmenter. Also thanks to @csinva and @afnanmmir for the reviews! I am merging this now. @nick11roberts, please create a separate PR to add the keywords for your PR - I will merge it directly!

Thank you @kaustubhdhole, and my apologies for the delay on the keywords! Please see https://github.com/GEM-benchmark/NL-Augmenter/pull/299 for the newly added keywords.

**Pull request merged**